DataFrame: convert from row based to a columnar value format

ryantxu opened the pull request:
Alternative to #18312.  

Key changes:
1. Rather than having a 'rows' object, each Field has a values interface
2. DataFrame and Field objects, now have properties that do not automatically read/write from JSON.
3. Added FieldDTO and DataFrameDTO that only include simple json objects
4. DataFrame helper is a new class that helps manage a DataFrame
5. FieldCache is now part of DataFrameHelper
6. DataFrameView<T = any> exposes the contents of DataFrame as any object

I am not wild about the names:
* DataFrameHelper
Any suggestions/alternatives would be great :)

After this is merged, I want to:
* cleanup the DisplayField / DisplayProcessor lifecycle

ryantxu committed:
add vector class

ryantxu committed:
move schema to display

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar

* grafana/master:
  TablePanel: Remove scroll option on TablePanel (#18318)
  Keyboard Shortcuts: Sign in to enable them (#18271)
  GitHub Templates: Pull Request Template update (#18300)
  Auth Proxy: Include additional headers as part of the cache key (#18298)
  grafana/toolkit: support windows paths (#18306)
  Chore: noImplicitAny Sub 500 errors (#18287)

ryantxu committed:
adding dataFrame helper

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar

* grafana/master:
  Fix failing end to end tests job for release (#18323)
  Fix OAuth error due to SameSite cookie policy (#18332)
  Chore: noImplictAny no errors left (#18303)
  [Shortcuts] Fixes shortcuts for moving time range backwards and forwards (#18305)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar

* grafana/master:
  Refactor: use data rather than series in stream callback(#18126)
  Keybindings: Disable on login url (#18331)

ryantxu committed:
merge master

ryantxu committed:
csv parsing

ryantxu committed:
almost passing tests

ryantxu committed:
tests almost pass

ryantxu committed:
JSON vs DataFrame

ryantxu committed:
merge master

ryantxu committed:
merge master

ryantxu committed:
use settings from fied

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  MSSQL: Change connectionstring to URL format to fix using passwords with semicolon (#18384)
  CloudWatch: Fix alerting for queries with Id (using GetMetricData) (#17899)

ryantxu committed:
merge master

ryantxu committed:
getLength() to .length

ryantxu commented:
@torkelo -- can you take a look and give general feedback?  I know it is a lot to grock, but I think this will help us significantly with any frontend data manipulation and allow for things like simple math and transformations

ryantxu committed:
merge master

dprokop commented:
@ryantxu ref `In messing with apache arrow and starting to work on some transformations, it seems worth considering using a column based structure rather than a row based structure.`. How is the column based structure better? I guess this is some obvious thing but would be good to have some kind documentation/explanation about your reasoning behind this as this concept seems to be of a fundamental kind :)

torkelo commented:
Spent half the day on fixing critical issues in 6.3.0 and getting 6.3.1 & 6.3.2 out, and then rest of the day was meetings & interviews. 

@dprokop so the main benefit is being able to remove/add columns without modifying coping a big rows array. 


Current structure: 
fields: [ { type: 'time', name: 'time' }, { type: 'value', name: 'value' } ]
rows[][]

where rows is a 2 dimensional array with an array for each row and each row has an array with same length as number of fields. 

So adding a new field or removing a field requires the whole 2d rows array to be rebuilt. 

In contract compare it to:

```ts
const frame: DataFrameJSON = {
    fields: [
      { name: 'time', type: FieldType.time, buffer: [100, 200, 300] },
      { name: 'name', type: FieldType.string, buffer: ['a', 'b', 'c'] },
      { name: 'value', type: FieldType.number, buffer: [1, 2, 3] },
    ],
  };
```

Here the values for each field (ie column) is grouped on the actual field. So with this, we can create a new with removed or added fields without modifying the original data structure or creating new copies of the preserved data. 

https://en.wikipedia.org/wiki/Column-oriented_DBMS

ryantxu commented:
This PR makes two key changes:
1. Restructure data so it is attached to the columns
2. Changes the API to access values from simple T[], to an interface [Vector<T=any>](https://github.com/grafana/grafana/pull/18391/files#diff-2d83db17b224e5df2fce07cb5ee79150R56)

These two changes will greatly simplify result manipulation.  Without needing to copy data into a new any[][], we can:
1. Select some fields from an existing DataFrame.  We could have a single query to get all data for a dashboard, and each visualisation just picks the fields it needs
2. "Join" fields into one DataFrame if there is a common Field
3. Use an abstraction to provide a variation of the data in a column.  For example the Table view in Explore has three copies of the date -- it could be one copy, that gets formatted on demand for the cells that are visible
4. Add computed columns -- See [ScaledVector](https://github.com/grafana/grafana/pull/18391/files#diff-5164ddf6f82b680634a4f39e55a177f3R43), this could be simple math, or functions

As is, Grafana usually asks people change their queries and datasource so results are structured for the visualisation.  I would like to be able to use/re-use almost any result in any visualisation.  In particular a 'stream' of incoming data that gets transformed and reused across a dashboard.

This will make some interesting UX options possible.  I really like how https://app.rawgraphs.io/ has you map data "Dimensions" to what the each visualisation supports:
![image](https://user-images.githubusercontent.com/705951/62755898-a1f39a00-ba2a-11e9-9a3d-5156589ca08c.png)



The disadvantages I see are:
* The DataFrame with Vector<T> can not automatically read/write JSON
* More complex to add/remove rows
* Sorting/searching is more complex
* Major API change

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  DataLinks: Apply scoped variables correctly (#18454)
  DataLinks: Use datapoint timestamp correctly when interpolating variables (#18459)
  API: Minor fix for team creation endpoint when using API key (#18252)
  Login: Adjust space between skip and its icon (#18407)
  Docs: Update Auth Proxy documentation (#18444)
  Docs: Minor Readme update (#18438)

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  Explore: Fix loading error for empty queries (#18488)
  Fix: Fixes stripping of $d in Explore urls (#18480)
  grafana/ui: fix toTimeTicks error (#18448)
  Docs: Adds details to Pull Request Checklist (#18471)
  DataLinks: respect timezone when displaying datapoint's timestamp in graph context menu (#18461)
  Chore: strictNullChecks, ColoringEditor and time_region_manager (#18442)
  Backend: Do not set SameSite cookie attribute if cookie_samesite is none (#18462)

ryantxu committed:
implement sort

ryantxu committed:
add a circular buffer

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  Fix: failing build after React bump (#18514)
  strictNullChecks: First batch (#18390)
  Chore: bump React to 16.9.0 (#18502)
  Docs: Adds a new security section (#18508)
  Docs: Update issue triage doc with external PRs section (#18464)
  Typo: fix typo in processDataFrame.ts comment (#18492)

torkelo commented on packages/grafana-ui/src/utils/flotPairs.ts:
```diff
@@ -1,24 +1,28 @@
-    const x = rows[i][xIndex];
-    let y = rows[i][yIndex];
+  for (let i = 0; i < length; i++) {
+    const x = vX.get(i);
```
Is it possible (as a future optimization) use indexers in cases like the above?

ryantxu replied:
what do you mean?

`vX[i]`?  that is possible, but requires proxy if the implementation is not a real array

torkelo replied:
Just what performance optimizations we could do in the future to make transformations and loops like this more performant and not require method calls

torkelo reviewed (commented):
This looks very promising. 

What is left todo in this PR? 

Regarding names I had some alternative suggestions in line (not sure they are better). 

* DataFrameJSON
* DataFrameDTO

* DataFrameHelper
* DataFrameWrapper 
* DataFrameAccessor

Not sure which is best,  DataFrameHelper might still be best. 

Bugs
* The streaming works differently in this branch, the time range is fixed to a just 3 points or so Graph time range flies past
* The Gauge & Bar gauge is static in streaming with Last calc (ie no value update)

ryantxu commented on packages/grafana-data/src/utils/dataFrameHelper.ts:
```diff
@@ -0,0 +1,308 @@
+   *
+   * TODO: really just want a circular buffer
+   */
+  slice(start?: number, end?: number) {
```
It is used by the streaming clients -- they add data at the end of the list and then remove the first rows to keep the buffer the same size.

I want to remove this and use `CircularVector` -- essentially allocate an array and overwrite values

ryantxu replied:
I removed it

ryantxu commented on packages/grafana-data/src/types/dataFrame.ts:
```diff
@@ -0,0 +1,99 @@
+  noValue?: string;
+}
+
+export interface FieldJSON<T = any> {
```
👍

ryantxu commented:
What is left todo in this PR?  hopefully not much :)
* I will replace the `slice` stuff with `CircularBuffer` and test streaming.
* I will rename Field/DataFrameJSON to Field/DataFrameDTO

DataFrameHelper name?  Still not sure, I think the best candidates are:
 * DataFrameHelper
 * DataFrameWrapper
 * MutableDataFrame  
 * DataFrameBuilder

ryantxu committed:
merge master

ryantxu committed:
FieldJSON to FieldDTO

ryantxu committed:
Circular vector in streams

ryantxu commented:
@hugohaggmark -- can you take a look at this?  I merged in your changes from #18475, and am not totally sure what and how to check the explore changes :)

After this is merged, I think we can greatly simplify the loki >>> explorer data lifecycle

ryantxu committed:
fix InputDataSource

ryantxu committed:
remove createField utility

ryantxu committed:
move dto

ryantxu committed:
clear cached stats for streaming

ryantxu commented:
@torkelo -- I made some more changes since you looked:
* XxxJSON -> XxxDTO
* removed the slice methods and the streaming stuff works with CircularVectors.  this needs more work/cleanup, but should be independent.
* I change the signature of DTO vs regular object to have the same field name `values` in DTO it is array, in the regular one it is Vector.  Previously it had buffer vs values.
* I fixed the streaming updates.  (caching stats works!) now it clears the cached stats.

Can you take another look?

hugohaggmark commented:
> @hugohaggmark -- can you take a look at this? I merged in your changes from #18475, and am not totally sure what and how to check the explore changes :)
> 
> After this is merged, I think we can greatly simplify the loki >>> explorer data lifecycle

Looking at it

hugohaggmark reviewed (commented):
I've looked into the Explore parts and done some manual testing and it almost everything works as expected.

- #18488 has not been merged in this branch as I frequently get `Cannot read property 'split' of undefined` when working in Explore
- The TestDatas `StreamingClient` stopped working for both `signal` and `logs` in Explore

ryantxu committed:
merge master

dprokop commented on packages/grafana-data/src/utils/dataFrameHelper.ts:
```diff
@@ -0,0 +1,264 @@
+    this.updateTypeIndex(field);
+
+    if (this.fieldByName[field.name]) {
+      console.warn('Duplicate field names in DataFrame: ', field.name);
```
I wonder if we shouldn't throw here instead of warning. `getFieldByName` becomes not predictable unless you know the implementation.

ryantxu replied:
I don't think we can throw an exception since there is no existing contract that says field (columns) can not have the same name.

I do think it is OK to say "You will have unpredictable behavior *if* fields have the same name"

ryantxu replied:
Just added https://github.com/grafana/grafana/pull/18391/commits/3292fb717692c400c8b34765a446ad2155edfe44 this makes the behavior more clear (i hope)

dprokop reviewed (commented):
Quite a lot to process @ryantxu but it looks good! My feeling is, when compared with the rows approach, that this is way easier to wrap your head around and reason about the data. 

I think this work also deserves some writeup as blog post as there is a lot of reasoning that might be interesting for the community!

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  CI: Fail build if yarn.lock is not up to date (#18555)
  Chore: Updates react-dependant packages to address react warnings (#18549)
  Prometheus: Fix regression of rerunning query on legend/interval change (#18147)
  Explore/Prometheus: More consistently allows for multi-line queries (#18362)
  Login: Fixes undefined redirect (#18545)

ryantxu committed:
make duplicate name constraints more clear

ryantxu committed:
use full results, not delta

ryantxu committed:
move explore delta work to another PR

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  Toolkit: write PR report to a folder with the circle build number (#18560)

ryantxu committed:
migrate rows based DataFrame to dataframe

ryantxu committed:
move DataFrameView to its own file and better property enumeration

ryantxu committed:
update comment

ryantxu committed:
add back toArray()

hugohaggmark commented:
This solves the streaming part in Explore for TestData for me. I talked to @torkelo and `delta` is needed for the Prometheus and Live streaming case. Don't know if `delta` is bad choice of naming for what its used for though.

```ts
appendRows(append: any[][]) {
    // Trim the maximum row count
    const { stream, values, data } = this;

    // Append all rows
    const delta = new DataFrameHelper({
      fields: data.fields.map(field => {
        const { config, name, type } = field;
        return {
          config,
          name,
          type,
          values: [],
        };
      }),
    });
    for (let i = 0; i < append.length; i++) {
      const row = append[i];
      delta.appendRow(row);
      for (let j = 0; j < values.length; j++) {
        values[j].append(row[j]); // Circular buffer will kick out old entries
      }
    }
    // Clear any cached values
    for (let j = 0; j < data.fields.length; j++) {
      data.fields[j].calcs = undefined;
    }
    stream.data = [data];
    stream.delta = [delta];

    // Broadcast the changes
    if (this.observer) {
      this.observer(stream);
    } else {
      console.log('StreamWorker working without any observer');
    }

    this.last = Date.now();
  }
```

torkelo reviewed (approved):
LGTM

what can go wrong? 🙈

ryantxu committed:
Merge remote-tracking branch 'grafana/master' into columnar2

* grafana/master:
  Packages: Temporarily skip canary releases if packages build fail (#18577)
  Update latest.json to latest stable version (#18575)
  Docs: Update changelog for v6.3.3 (#18569)
  Graph: Fixed issue clicking on series line icon (#18563)
  grafana/toolkit: Unpublish previous "next" version when releasing a new one (#18552)

ryantxu commented:
> what can go wrong? 🙈

I genuinely burst out laughing when I read this
